plugins/node/opentelemetry-instrumentation-awslambda/src/awslambda.ts

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as path from 'path';
+
+import {
+  InstrumentationBase,
+  InstrumentationNodeModuleDefinition,
+  InstrumentationNodeModuleFile,
+  isWrapped,
+  safeExecuteInTheMiddle,
+} from '@opentelemetry/instrumentation';
+import { diag, Span, SpanKind, SpanStatusCode } from '@opentelemetry/api';
+import { FaasAttribute } from '@opentelemetry/semantic-conventions';
+
+import { Callback, Context, Handler } from 'aws-lambda';
+
+import { LambdaModule } from './types';
+import { VERSION } from './version';
+
+export class AwsLambdaInstrumentation extends InstrumentationBase {
+  constructor() {
+    super('@opentelemetry/instrumentation-awslambda', VERSION);
+  }
+
+  init() {
+    // _HANDLER and LAMBDA_TASK_ROOT are always defined in Lambda.
+    const taskRoot = process.env.LAMBDA_TASK_ROOT!;
+    const handlerDef = process.env._HANDLER!;
+
+    const handler = path.basename(handlerDef);
+    const moduleRoot = handlerDef.substr(0, handlerDef.length - handler.length);
+
+    const [module, functionName] = handler.split('.', 2);
+
+    // Lambda loads user function using an absolute path.
+    let filename = path.resolve(taskRoot, moduleRoot, module);
+    if (!filename.endsWith('.js')) {
+      // Patching infrastructure currently requires a filename when requiring with an absolute path.
+      filename += '.js';
+    }
+
+    return [
+      new InstrumentationNodeModuleDefinition(
+        // NB: The patching infrastructure seems to match names backwards, this must be the filename, while
```

Comment:
I found this strange pattern that did allow patching the lambda handler.

Reply:
I'm not sure what you meant by matching name backwards ? It matches absolute path (https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-instrumentation/src/platform/node/instrumentation.ts#L107)

Reply:
Yeah, I mean that we pass the filename to a class called `InstrumentationNodeModuleDefinition`, but pass a module name to a class called `InstrumentationNodeModuleFile`. I'd expect file to refer to a filename :)

Reply:
Yeah thats weird, i'm not sure if it's because of the resolving done by lamda. With other instrumentation we do the opposite (module is a name and file is an relative path to the module): https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts#L59

Reply:
this indeed looks strange as it should be a module name if user requires for example `http` then the module name is `http` so in this case it should be `aws-lambda` not a file name. If you need to patch some additional files then this is what `InstrumentationNodeModuleFile` is for